truss-chains/truss_chains/code_gen.py

```diff
@@ -511,13 +514,74 @@ def _gen_truss_chainlet_file(
 # Truss Gen ############################################################################
 
 
-def _copy_python_source_files(root_dir: pathlib.Path, dest_dir: pathlib.Path) -> None:
+def _copy_python_source_files(
+    root_dir: pathlib.Path,
+    dest_dir: pathlib.Path,
+    max_files: int = 1000,
+    max_size_mb: int = 100,
+    time_limit_seconds: int = 3,
+) -> None:
     """Copy all python files under root recursively, but skips pycache."""
 
-    def python_files_only(_, names):
-        return [name for name in names if name == "__pycache__"]
+    def exclude_filter(_, names):
+        return [name for name in names if name in ("__pycache__", ".mypy_cache")]
 
-    shutil.copytree(root_dir, dest_dir, ignore=python_files_only, dirs_exist_ok=True)
+    start_time = time.time()
+    total_size = 0
+    total_files = 0
+    max_size_bytes = max_size_mb * 1024 * 1024
+    limit_issue = False
+
+    for dir_path, dir_names, filenames in os.walk(root_dir):
+        # In-place modification affects `os.walk` iterator intentionally.
+        dir_names[:] = [
+            d for d in dir_names if d not in exclude_filter(dir_path, dir_names)
+        ]
+        filenames = [
+            f for f in filenames if f not in exclude_filter(dir_path, filenames)
+        ]
+        for filename in filenames:
+            total_files += 1
+            filepath = pathlib.Path(dir_path) / filename
+            total_size += filepath.stat().st_size
+            if total_files > max_files or total_size > max_size_bytes:
+                limit_issue = True
+                break
+            if (time.time() - start_time) > time_limit_seconds:
+                limit_issue = True
+                break
+
+        if limit_issue:
+            break
+
+    if limit_issue:
+        if total_files > max_files:
+            issue_msg = f"found a large number of files (>{max_files})"
+        elif total_size > max_size_bytes:
+            issue_msg = (
+                f"found a large amount of data (>{total_size / (1024 * 1024):.2f} MB)"
+            )
+        else:
+            issue_msg = f"it took longer than {time_limit_seconds} seconds to scan"
+
+        msg = (
+            f"All files nested in the chains workspace directory are included in the "
+            "remote deployment. To prevent accidental inclusion of more files than "
+            "intended we scanned \n"
```

Comment:
could we list out explicitly which files will be copied? this would make it very clear to the users whether they've structured their chain correctly

Reply:
Hmm, that might be a wall of files, not sure if that's really so useful to show by default. I'll skip that for now.